subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c

```diff
@@ -0,0 +1,1251 @@
+
+	/* Set flag that ICMsg is bounded and now, endpoint bounding may start. */
+	atomic_or(&dev_data->flags, FLAG_ICMSG_BOUNDED);
+	schedule_ept_bound_process(dev_data);
+}
+
+/**
+ * Open the backend instance callback.
+ */
+static int open(const struct device *instance)
+{
+	const struct icmsg_with_buf_config *conf = instance->config;
+	struct backend_data *dev_data = instance->data;
+
+	static const struct ipc_service_cb cb = {
+		.bound = bound,
+		.received = received,
+		.error = NULL,
+	};
+
+	LOG_DBG("Open instance 0x%08X, initiator=%d", (uint32_t)instance,
+		dev_data->is_initiator ? 1 : 0);
+	LOG_DBG("  ICMsg, TX %d at 0x%08X, RX %d at 0x%08X",
+		(uint32_t)conf->icmsg_config.tx_shm_size,
+		(uint32_t)conf->icmsg_config.tx_shm_addr,
+		(uint32_t)conf->icmsg_config.tx_shm_size,
+		(uint32_t)conf->icmsg_config.tx_shm_addr);
+	LOG_DBG("  TX %d blocks of %d bytes at 0x%08X, max allocable %d bytes",
+		(uint32_t)conf->tx.block_count,
+		(uint32_t)conf->tx.block_size,
+		(uint32_t)conf->tx.blocks_ptr,
+		(uint32_t)(conf->tx.block_size * conf->tx.block_count -
+			   BLOCK_HEADER_SIZE));
+	LOG_DBG("  RX %d blocks of %d bytes at 0x%08X, max allocable %d bytes",
+		(uint32_t)conf->rx.block_count,
+		(uint32_t)conf->rx.block_size,
+		(uint32_t)conf->rx.blocks_ptr,
+		(uint32_t)(conf->rx.block_size * conf->tx.block_count -
+			   BLOCK_HEADER_SIZE));
+
+	return icmsg_open(&conf->icmsg_config, &dev_data->icmsg_data, &cb,
+			  (void *)instance);
+}
+
+/**
+ * Endpoint send callback function (with copy).
+ */
+static int send(const struct device *instance, void *token, const void *msg, size_t len)
+{
+	struct backend_data *dev_data = instance->data;
+	struct ept_data *ept = token;
+	uint32_t alloc_size;
+	uint8_t *buffer;
+	int r;
+
+	/* Allocate the buffer. */
+	alloc_size = len;
+	r = alloc_tx_buffer(dev_data, &alloc_size, &buffer, K_FOREVER);
+	if (r < 0) {
+		return r;
+	}
+
+	/* Copy data to allocated buffer. */
+	memcpy(buffer, msg, len);
+
+	/* Send data message. */
+	return send_block(dev_data, MSG_DATA, ept->addr, r, len);
+}
+
+/**
+ * Backend endpoint registration callback.
+ */
+static int register_ept(const struct device *instance, void **token,
+			const struct ipc_ept_cfg *cfg)
+{
+	struct backend_data *dev_data = instance->data;
+	struct ept_data *ept = NULL;
+	int ept_index;
+	int r = 0;
+
+	/* Reserve new endpoint index. */
+	ept_index = atomic_inc(&dev_data->flags) & FLAG_EPT_COUNT_MASK;
+	if (ept_index >= NUM_EPT) {
+		LOG_ERR("Too many endpoints");
+		__ASSERT_NO_MSG(false);
+		return -ENOMEM;
+	}
+
+	/* Add new endpoint. */
+	ept = &dev_data->ept[ept_index];
+	ept->cfg = cfg;
+	if (dev_data->is_initiator) {
+		ept->addr = ept_index;
+		dev_data->ept_map[ept->addr] = ept->addr;
+	}
+	atomic_set(&ept->state, EPT_CONFIGURED);
+
+	/* Keep endpoint address in token. */
+	*token = ept;
+
+	/* Rest of the bounding will be done in the system workqueue. */
+	schedule_ept_bound_process(dev_data);
```

Comment:
Is it needed on no initiator side?

Reply:
Yes

Reply:
Could we just assume that endpoint registration is done after icmsg control channel is open. It could simplify logic because we wouldn't need to have states and bound message could be sent directly from this function.